xls/modules/zstd/zstd_dec_test.cc

```diff
@@ -0,0 +1,297 @@
+      for (const auto& tm : events.trace_msgs) {
+        std::cout << "[TRACE] " << tm.message << "\n";
+      }
+    }
+  }
+
+  const char* proc_name = "__zstd_dec__ZstdDecoderTest_0_next";
+  const char* recv_channel_name = "zstd_dec__output_s";
+  const char* send_channel_name = "zstd_dec__input_r";
+
+  const char* ir_file = "xls/modules/zstd/zstd_dec_test.ir";
+
+  std::unique_ptr<Package> package;
+  std::unique_ptr<SerialProcRuntime> interpreter;
+  ChannelQueue *recv_queue, *send_queue;
+
+  void PrintVector(absl::Span<uint8_t> vec) {
+    for (int i = 0; i < vec.size(); i += 8) {
+      std::cout << "0x" << std::hex << std::setw(3) << std::left << i
+                << std::dec << ": ";
+      for (int j = 0; j < sizeof(uint64_t) && (i + j) < vec.size(); j++) {
+        std::cout << std::setfill('0') << std::setw(2) << std::hex
+                  << static_cast<unsigned int>(vec[i + j]) << std::dec << " ";
+      }
+      std::cout << "\n";
+    }
+  }
+
+  void DecompressWithLibZSTD(std::vector<uint8_t> encoded_frame,
+                             std::vector<uint8_t>& decoded_frame) {
+    size_t buff_out_size = ZSTD_DStreamOutSize();
+    uint8_t* const buff_out = new uint8_t[buff_out_size];
+
+    ZSTD_DCtx* const dctx = ZSTD_createDCtx();
+    EXPECT_FALSE(dctx == nullptr);
+
+    void* const frame = static_cast<void*>(encoded_frame.data());
+    size_t const frame_size = encoded_frame.size();
+    // Put the whole frame in the buffer
+    ZSTD_inBuffer input_buffer = {frame, frame_size, 0};
+
+    while (input_buffer.pos < input_buffer.size) {
+      ZSTD_outBuffer output_buffer = {buff_out, buff_out_size, 0};
+      size_t decomp_result =
+          ZSTD_decompressStream(dctx, &output_buffer, &input_buffer);
+      bool decomp_success = ZSTD_isError(decomp_result) != 0u;
+      EXPECT_FALSE(decomp_success);
+
+      // Append output buffer contents to output vector
+      decoded_frame.insert(
+          decoded_frame.end(), static_cast<uint8_t*>(output_buffer.dst),
+          (static_cast<uint8_t*>(output_buffer.dst) + output_buffer.pos));
+
+      EXPECT_TRUE(decomp_result == 0 && output_buffer.pos < output_buffer.size);
+    }
+
+    ZSTD_freeDCtx(dctx);
+    delete[] buff_out;
+  }
+
+  void ParseAndCompareWithZstd(std::vector<uint8_t> frame) {
+    std::vector<uint8_t> lib_decomp;
+    DecompressWithLibZSTD(frame, lib_decomp);
+    size_t lib_decomp_size = lib_decomp.size();
+    std::cout << "lib_decomp_size: " << lib_decomp_size << "\n";
+
+    std::vector<uint8_t> sim_decomp;
+    size_t sim_decomp_size_words =
+        (lib_decomp_size + sizeof(uint64_t) - 1) / sizeof(uint64_t);
+    size_t sim_decomp_size_bytes =
+        (lib_decomp_size + sizeof(uint64_t) - 1) * sizeof(uint64_t);
+    sim_decomp.reserve(sim_decomp_size_bytes);
+
+    // Send compressed frame to decoder simulation
+    for (int i = 0; i < frame.size(); i += 8) {
+      auto span = absl::MakeSpan(frame.data() + i, 8);
```

Comment:
Looks like this could access out-of-bounds memory if frame size is not a multiple of 8.